### PR TITLE
Feedback for the enumerable methods #1

### PR DESCRIPTION
**Pull request by thsvr**
Hi, 

Can you please check my code and give me feedback? 

Besides that, I let some tests in comments just in case I need to fix something but I will delete them after getting my project approved.

Thank you in advance :) 

Cheers, 

Thaís

### CODE REVIEW

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Hi @thsvr you have a very nice approach to this project, you implemented a lot of these functions well :clap:  please see the use cases left to fulfil below 

**my_each**
my_each returns an Enumerator if no block is given
my_each_with_index returns an enumerator if no block is given

**my_select**
my_select returns an enumerator if no block is given

**my_any**
my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

**my_none**
my_none? when no block or argument is given returns true only if none of the collection members is true
my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex
my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

**my_count**
my_count counts the number of items in enum that are equal to item if an argument is given
my_count counts the number of elements yielding a true value if a block is given

**my_map**
my_map returns an Enumerator if no block is given

**my_inject**
my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

Please try to make changes and request another review when you are done :pray:

**Comment by thsvr** [comment]
Hi,

I did the changes asked by Raphael, so can you please check if my project is able to be approved and then give feedback?

I let some tests in comments just in case I need to fix something again but I will delete them after getting my project approved and before merging it to developer/master branches.

Thank you in advance :)

Cheers,

Thaís

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+      x += 1
+    end
+    self
+  end
+
+  # my_each_with_index 2
+  def my_each_with_index
+    (0...size).each do |x|
+      yield self[x], x
+    end
```
Comment (samgaco) [inline enumerable_methods.rb]: ### 1.  Should return an enumerator when no block or arguments are given

expected:

```

irb(main):129:0> arr = [1,2,3]
=> [1, 2, 3]
irb(main):130:0> arr.each_with_index
=> #<Enumerator: [1, 2, 3]:each_with_index>
```

current method:

```
irb(main):131:0> arr.my_each_with_index
...
LocalJumpError (no block given (yield))
```

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+      my_each { |x| return false unless x == par }
+    end
+    true
+  end
+
+  # my_any? 5
+  def my_any?(_par = nil)
+    if block_given?
+      my_each { |x| return true if yield(x) }
+    elsif param.class == Class
```
Comment (samgaco) [inline enumerable_methods.rb]: ### 2. should return true if at least one of the elements of the passed collection of elements is not false or nil , example compared with the current one:

```
irb(main):132:0> arr = [false, false, true]
=> [false, false, true]
irb(main):133:0> arr.any?
=> true
irb(main):134:0> arr.my_any?
...
NameError (undefined local variable or method `param' for [false, false, true]:Array)
```

### 3. when passing a collection of elements, and a class as an argument on the method, it should evaluate whether at least one of those elements is a member of that class:

```
irb(main):142:0> arr = [1,2,"ooo"]
=> [1, 2, "ooo"]
irb(main):143:0> arr.any?(Integer)
=> true
irb(main):144:0> arr.any?(String)
=> true
```

### 4. Same as in the above case for regular expressions:

```
irb(main):169:0> arr = ["zz", 1]
=> ["zz", 1]
irb(main):170:0> arr.any?(/[a-z]/)
=> true
```

### 5. also when the input is a pattern:

```
irb(main):174:0> arr = ["zz", 1]
=> ["zz", 1]
irb(main):175:0> arr.any?(1)
=> true
irb(main):176:0> arr.any?("zz")
=> true
```

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      my_each { |x| return true if x == param }
+    end
+    false
+  end
+
+  # my_none? 6
+  def my_none?(_par = nil)
+    if block_given?
+      my_each { |x| return false if yield(x) }
+    elsif param.class == Class
```
Comment (samgaco) [inline enumerable_methods.rb]: ### (6,7,8,9) same points as in 2,3,4,5 in **my_any?**

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @thsvr :high_brightness: 

Thanks for adding a comment with the changes done so far, it is helpful :+1: 

- [x] I left some comments with 9 points to be addressed before the approval. Points 6,7,8 and 9 are the same cases as in 2,3,4,5 in my_any, but this time tested for my_none, so they should look for the opposite result (checking that there is no element following a certain condition, a regular expression, a given pattern, or a given class)

Please :pray: add these behaviors and submit back for another code review. :muscle: 

----------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

**Comment by thsvr** [comment]
Hi,

I did the changes asked by Samuel and I guess now it works fine. 
So can you please check if my project is able to be approved and then give feedback?

I let some tests in comments just in case I need to fix something again but I will delete them after getting my project approved and before merging it to developer/master branches.

Thank you in advance :)

Cheers,

Thaís

**Review by Ispirett — APPROVED** [review]
Excellent work with the changes made @thsvr  you can merge now.

image

**status: Approved**

**Pull request merged**